source/agora/network/Manager.d

```diff
@@ -642,47 +643,60 @@ public class NetworkManager
             {
-                // Do not query the registry about ourself
-                if (key == this.validator_config.key_pair.address)
-                    continue;
+                this.quorum_set_keys.put(hash);
+                if (hash !in this.connected_validator_keys)
+                    this.required_peers.put(hash);
+            }
 
-                taskman.runTask
-                ({
-                    retry!
+            if (this.registry_client !is null && required_peer_utxos !is null)
+            {
+                foreach (hash, utxo; required_peer_utxos)
+                {
+                    auto key = utxo.output.address;
+                    // Do not query the registry about ourself
+                    if (key == this.validator_config.key_pair.address)
+                        continue;
+
+                    taskman.runTask
                     ({
-                        auto payload = this.registry_client.getValidator(key);
-                        if (payload == RegistryPayload.init)
-                        {
-                            log.warn("Could not find mapping in registry for key {}", key);
-                            return false;
-                        }
+                        retry!
+                        ({
+                            auto payload = this.registry_client.getValidator(key);
+                            if (payload == RegistryPayload.init)
+                            {
+                                log.warn("Could not find mapping in registry for key {}", key);
+                                return false;
+                            }
 
-                        if (!payload.verifySignature(key))
-                        {
-                            log.warn("RegistryPayload signature is incorrect for {}", key);
-                            return false;
-                        }
-                        foreach (addr; payload.data.addresses)
-                            this.addAddress(addr);
-                        return true;
-                    },
-                    )(taskman, 3, 2.seconds, "Exception happened while trying to get validator addresses");
-                });
+                            if (!payload.verifySignature(key))
+                            {
+                                log.warn("RegistryPayload signature is incorrect for {}", key);
+                                return false;
+                            }
+                            foreach (addr; payload.data.addresses)
+                                this.addAddress(addr);
+                            return true;
+                        },
+                        )(taskman, 3, 2.seconds, "Exception happened while trying to get validator addresses");
+                    });
+                }
             }
         }
+        catch (Exception ex)
+        {
+            log.error("Error discovering peers: {}", ex);
+        }
```

Comment:
Looks like the `try` clause is only because of the `foreach` ? If so, `byKeyValue`.

Reply:
Will update.